File: app/schemas/assignment.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class AssignmentBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: Optional[str] = Field(default=None, max_length=5000)
    due_date: Optional[datetime] = None
    max_score: int = Field(default=100, ge=1, le=1000)
    is_published: bool = False


class AssignmentCreate(AssignmentBase):
    course_id: int = Field(gt=0)


class AssignmentUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = Field(default=None, max_length=5000)
    due_date: Optional[datetime] = None
    max_score: Optional[int] = Field(default=None, ge=1, le=1000)
    is_published: Optional[bool] = None


class AssignmentPublishIn(BaseModel):
    is_published: bool = True


class AssignmentOut(AssignmentBase):
    id: int
    course_id: int
    school_id: int
    created_by: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SubmissionCreate(BaseModel):
    submission_text: Optional[str] = None
    attachment_url: Optional[str] = None


class SubmissionGrade(BaseModel):
    score: int = Field(ge=0)
    feedback: Optional[str] = None


class SubmissionOut(BaseModel):
    id: int
    assignment_id: int
    student_id: int
    submission_text: Optional[str] = None
    attachment_url: Optional[str] = None
    status: str
    score: Optional[int] = None
    feedback: Optional[str] = None
    graded_by: Optional[int] = None
    graded_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

File: mhike-school-web/lib/services/assignment.ts
import { apiDelete, apiGet, apiPatch, apiPost } from "@/lib/api";

export type AssignmentOut = {
    id: number;
    title: string;
    description?: string | null;
    course_id: number;
    school_id: number;
    created_by: number;
    due_date?: string | null;
    max_score: number;
    is_published: boolean;
    created_at: string;
};

export type AssignmentCreate = {
    title: string;
    description?: string | null;
    course_id: number;
    due_date?: string | null;
    max_score?: number;
    is_published?: boolean;
};

export type AssignmentUpdate = Partial<AssignmentCreate>;

export type AssignmentPublishIn = {
    is_published: boolean;
};

export type AssignmentSubmissionOut = {
    id: number;
    assignment_id: number;
    student_id: number;
    school_id: number;
    submission_text?: string | null;
    attachment_url?: string | null;
    status: string;
    score?: number | null;
    feedback?: string | null;
    graded_by?: number | null;
    graded_at?: string | null;
    submitted_at: string;
    created_at: string;
};

export type AssignmentSubmissionSubmit = {
    submission_text?: string | null;
    attachment_url?: string | null;
};

export type AssignmentSubmissionGrade = {
    score?: number | null;
    feedback?: string | null;
    status?: string;
};

/* =========================
   Assignments
========================= */

export async function createAssignment(
    body: AssignmentCreate,
): Promise<AssignmentOut> {
    return apiPost<AssignmentOut>("/assignments", {
        ...body,
        max_score: body.max_score ?? 100,
        is_published: body.is_published ?? false,
    });
}

export async function getMyTeacherAssignments(): Promise<AssignmentOut[]> {
    return apiGet<AssignmentOut[]>("/assignments/me");
}

export async function getMyStudentAssignments(): Promise<AssignmentOut[]> {
    return apiGet<AssignmentOut[]>("/assignments/my");
}

export async function getAssignment(
    assignmentId: number,
): Promise<AssignmentOut> {
    return apiGet<AssignmentOut>(`/assignments/${assignmentId}`);
}

export async function updateAssignment(
    assignmentId: number,
    body: AssignmentUpdate,
): Promise<AssignmentOut> {
    return apiPatch<AssignmentOut>(`/assignments/${assignmentId}`, body);
}

export async function publishAssignment(
    assignmentId: number,
    body: AssignmentPublishIn = { is_published: true },
): Promise<AssignmentOut> {
    return apiPost<AssignmentOut>(
        `/assignments/${assignmentId}/publish`,
        body,
    );
}

export async function unpublishAssignment(
    assignmentId: number,
): Promise<AssignmentOut> {
    return publishAssignment(assignmentId, { is_published: false });
}

export async function deleteAssignment(
    assignmentId: number,
): Promise<void> {
    return apiDelete<void>(`/assignments/${assignmentId}`);
}

/* =========================
   Submissions
========================= */

export async function submitAssignment(
    assignmentId: number,
    body: AssignmentSubmissionSubmit,
): Promise<AssignmentSubmissionOut> {
    return apiPost<AssignmentSubmissionOut>(
        `/assignment-submissions/${assignmentId}/submit`,
        body,
    );
}

export async function getMySubmission(
    assignmentId: number,
): Promise<AssignmentSubmissionOut> {
    return apiGet<AssignmentSubmissionOut>(
        `/assignment-submissions/${assignmentId}/me`,
    );
}

export async function listAssignmentSubmissions(
    assignmentId: number,
): Promise<AssignmentSubmissionOut[]> {
    return apiGet<AssignmentSubmissionOut[]>(
        `/assignment-submissions/assignment/${assignmentId}`,
    );
}

export async function gradeSubmission(
    submissionId: number,
    body: AssignmentSubmissionGrade,
): Promise<AssignmentSubmissionOut> {
    return apiPost<AssignmentSubmissionOut>(
        `/assignment-submissions/${submissionId}/grade`,
        body,
    );
}